### ISSUE TITLE
needs-ctx vars don't get ctx when used as function argument

### ISSUE THREAD

**Issue by bobisageek** [borkdude/sci#600]
**version**

master branch

**platform**

JVM

**problem**

In the work on "varify"ing some of the core vars, I stumbled across a difference in behavior between a function and a var with the value of a function. When a var is created, it seems to 'break' the behavior of the `needs-ctx` metadata when that var is used as an argument to a HOF - specifically, we get an ArityException that seems to indicate that the context is not being passed to the invoke call.

**repro**

https://github.com/borkdude/sci/blob/50ce03d3223658bfd8e30c0779afd2834b713343/src/sci/impl/namespaces.cljc#L1001

`core-var` find-ns

```clj
; works when in function position
(eval-string "(find-ns 'clojure.set)")
=> #object[sci.impl.vars.SciNamespace 0x206fb897 "clojure.set"]

; errors when function argument
(eval-string "(mapv find-ns ['clojure.set])")
Execution error (ArityException) at sci.impl.vars.SciVar/null (vars.cljc:323).
Wrong number of args (1) passed to: sci.impl.namespaces/sci-find-ns
```

and this behavior isn't exhibited with functions with the metadata attached
https://github.com/borkdude/sci/blob/50ce03d3223658bfd8e30c0779afd2834b713343/src/sci/impl/namespaces.cljc#L1003

```clj
(eval-string "(mapv find-var ['clojure.core/set])")
=> [#'clojure.core/set]
```

**expected behavior**

```clj
(eval-string "(mapv find-ns ['clojure.set])")
=> [#object[sci.impl.vars.SciNamespace 0x206fb897 "clojure.set"]]
```

I've been trying to dig into the details, but I'm a bit stuck. I can see that in function position, both the var and the function end up going through `return-needs-ctx-call` (as I'd expect), but when used as an argument, it seems like neither the var nor the function go through needs-ctx in the analyzer.

It looks like if I change the var creation (i.e. `core-var) to just do a `with-meta` instead of making a new var, that resolves the symptom, and that should be sufficient for adding the namespace meta. I'm not sure if it's worth doing more work to address the var behavior, but I wanted to get your thoughts on it before adding namespace meta to more of core.

**Comment by borkdude** [comment]
Just wondering: did you encounter this breaking behavior in a test, or just by luck?

**Comment by bobisageek** [comment]
in a test:
https://github.com/borkdude/sci/blob/50ce03d3223658bfd8e30c0779afd2834b713343/test/sci/namespaces_test.cljc#L89-L90

When I put the existing version of `core-var` on `alias`, this test started failing

**Comment by borkdude** [comment]
@bobisageek OK, I'd say you go ahead with the var-izing and make a PR and I will merge this to a branch and fix that issue. OR you can look at the issue yourself, I'm pretty sure it's something in the `analyzer.cljc` namespace which looks at the metadata but not if it's a var or so.

**Comment by borkdude** [comment]
@bobisageek Think I found it. We might have to switch around these conditions:

https://github.com/borkdude/sci/blob/50ce03d3223658bfd8e30c0779afd2834b713343/src/sci/impl/analyzer.cljc#L1282-L1290

**Comment by bobisageek** [comment]
fixed in #601
